### PR TITLE
Encryption: Stop decrypting EE encrypted secrets with legacy encryption

### PR DESCRIPTION
**Pull request by undef1nd**
**What this PR does / why we need it**:
Decrypting EE secrets with legacy encryption don't fail, it produces unexpected output. Of course, it is not expected to switch EE feature on and off without prior rolling back secrets via CLI command (rolling back from EE to legacy). 

I think it's better to explicitly return an error upon attempt to decrypt EE secret with legacy encryption, than silently produce a wrong output.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
**_Question_**: Do I need to backport it to any version < v9.x?

### CODE REVIEW

File: pkg/services/secrets/manager/manager.go [hunk 1]
```diff
@@ -291,8 +291,15 @@ func newRandomDataKey() ([]byte, error) {
 }
 
 func (s *SecretsService) Decrypt(ctx context.Context, payload []byte) ([]byte, error) {
+	if len(payload) == 0 {
+		return nil, fmt.Errorf("unable to decrypt empty payload")
+	}
+
 	// Use legacy encryption service if featuremgmt.FlagDisableEnvelopeEncryption toggle is on
 	if s.features.IsEnabled(featuremgmt.FlagDisableEnvelopeEncryption) {
+		if len(payload) > 0 && payload[0] == '#' {
```
Comment (joanlopez) [inline pkg/services/secrets/manager/manager.go:300]: I really like the idea of detecting & preventing an EE encrypted secret to be decrypted with legacy encryption 🔝 

I only have question about that:
- Can we guarantee secrets NOT encrypted with EE will never starts with '#'? 

I'm thinking that, alternatively, maybe we could try to add a bit more restrictive check, although that would be a bit more risky.

Additionally:
- Do you think it'd make sense to backport these changes to `v8.5`?


I think we've been able to reproduce the same issue (EE encrypted secrets decrypted with legacy encryption without apparent error) in v8.5 so, maybe it worths adding that check in the latest v8.5 🤷🏻  Any thoughts?

Thanks!

Comment (undef1nd) [reply pkg/services/secrets/manager/manager.go:300]: >Can we guarantee secrets NOT encrypted with EE will never starts with '#'?

Given that we identify secret as the one encrypted by EE based on its first byte being '#', I think it's safe to assume none of legacy secrets should start with this sign.

Comment (undef1nd) [reply pkg/services/secrets/manager/manager.go:300]: Also, if I get everything right, legacy have salt prepended to the ciphertext. Looking [at this code](https://github.com/grafana/grafana/blob/main/pkg/util/encoding.go#L19) the salt can't contain '#' as it's generated from alphanumeric characters.

@sakjur please, correct me if I'm wrong.

Comment (joanlopez) [reply pkg/services/secrets/manager/manager.go:300]: Amazing, then 🚀

**Review by joanlopez — APPROVED** [review]

**Review by sakjur — APPROVED** [review]

**Pull request merged**